### ISSUE TITLE
Kinvey discussion

### ISSUE THREAD

**Issue by relatedcode** [relatedcode/ParseAlternatives#39]

**Comment by mikkokam** [comment]
I've been testing Kinvey as a Parse replacement for a few days now.
Seems to have a lot of that "Parse feel" to it.
**Documentation** seems very good - something that can not be said about some others in your excellent list of alternatives.

Seems they have most of the **features** needed

- Data
- Users (+social identities)
- REST + JS + Android + iOS ( + many more)
- Push
- Analytics
- Offline (and caching) support / better than Parse had, I think -- but I'd code my own anyhow to make it work just as I like it to
- Server side code

Missing (a big minus): **no hosting**. So a web app needs to host the files elsewhere. Luckily, CORS at Kinvey seems to be set up properly, so moving Parse web apps to Kinvey should not be too difficult.

They do have a lot of features Parse does not. Just after the Parse experiences, I am not that sure I'd use the most exotic ones, though. Who knows which giant buys Kinvey next, only to kill it after a while.
So, might as well stick to a very clearly layered model in the front end code, separating the backend client code to a module with standard functionality only. Currently have higher level Angular services for business logic in front end (for controllers to call), and these services call a DB service with CRUD verbs only. Keeping the DB client layer very dumb; it is free to implement very basic operations (REST is usually enough) without any knowledge of what it is actually doing.

Current Kinvey **pricing** is reasonable; there is a free tier for development and it is per app, not per developer, so it works exactly as it should. 
It's based on monthly active users (MAU) and $200 gets you 100 000 users, no limit on requests. Data storage is limited to 30 GB, which in many cases might limit the $200 plan earlier than the MAU does.
There is a steep upgrade after the 100k MAU / 30 GB -- the next plan is $2k a month - something to keep in mind if you get stuck with 100 001 users or 31 GB for a long time with no growth...

One point I was a bit worried about with Kinvey (as with many of the other alternatives, too): at Github, there is not a lot going on for the SDK's -- yes, a lot of updates, but very little stars, just 4 developers, no issues (?).

If anyone has experience working for a longer time with the Kinvey mBaaS, please -- do write here. Would be excellent to learn more about them.

**Comment by thelubin** [comment]
I've been reading about a lot of the different mBaaS alternatives. To me, Kinvey seems like a good option. Seems to have the same ease and feel as Parse. I will try and migrate my existing project to Kinvey and share my experiences. I would love to hear from people who have used it.

**Comment by Bizzi-Body** [comment]
Thanks @mikkokam - Your review is excellent.  I've been reading a lot about Kinvey and there is a lot to like about it.

As you mention, the big problem is that who knows what will happen to Kinvey in the future.  No one right now.  Your comments about layering your functionality is exactly the right thing to do.  Create interfaces that you call and in turn connect to Kinvey - if the worse happens then you've got easily identifiable endpoints that will need addressing.

I think I'm going to give them a try.....

**Comment by Bizzi-Body** [comment]
And already I am on the verge of giving up.  :(

I was attempting iOS/Swift.

Poor documentation and near nonexistent examples - none of which build.

Poor community forum - questions being asked - no one responding.  :(

**Comment by thelubin** [comment]
@Bizzi-Body That's sad to hear. Implementing log in/out and saving/deleting records worked fine for me. Haven't tried more stuff. But I agree, the documentation is sparse.

**Comment by Bizzi-Body** [comment]
I got anonymous user creation and logout working.  But I just could not get data out of the darn thing.  

I'm convinced that the problems I'm having are down to Xcode project set up.  Do you have an empty project you could share?  So that I can compare?

**Comment by thelubin** [comment]
Sorry, I'm on the run right now so I can't post a project. I followed this guide: http://devcenter.kinvey.com/ios/guides/getting-started. Did it manually.

**Comment by Bizzi-Body** [comment]
That's the Tut I started from, and I did it manually and using Pods.

**Comment by Bizzi-Body** [comment]
Right.  I had another go.  Clear head, fresh coffee etc etc.  And I have to say it worked just as it should.  I was able to very quickly create a simple table driven CRUD app.  I did also try the offline/caching and the few bits I attempted were okay, although I think really you'd want to back anything offline into CoreData or something like Realm.

So after spending a day playing..... 2 things stuck out......

a] Support - my question regarding how to create a "contains" query remains unanswered - I have a feeling that it is not possible using Swift without breaking out to ObjC.  I asked the question a couple of days ago - that no one has offered any response suggests either no one is listening or no one is using. I also notice that there has been only 1 other comment posted between then and now and that was to an already existing question - mine remains the last asked question.

b] Speed - my perception was that creating and updating records did not complete as quickly as Parse.  If I were to think about using Kinvey then I would certainly look to generate some metrics around this.  Of course I recognise that it could just be that I'm in "development" - so likely/might be nothing.

**Comment by DemonSkye** [comment]
Good morning every,

My name is Damien and I'm the Technical Support Team lead with Kinvey.   I apologize for my slow response time on your forum issue Bizzi-Body.  I've been in charge of handling your issue and it slipped my mind to transfer it to another member of my team, I've been out sick since Monday with either strep throat or the flu (I just got back from the doctor's a few moments ago).

http://devcenter.kinvey.com/ios/guides/datastore#operators -- The operator KCSIn is the one that you're after which is listed in the documentation with this example:

KCSQuery(onField: "colors", usingConditional: .KCSIn, forValue: ["orange", "black"])

B.  With regards to speed, we are running on the Mean stack and in my experience simple inserts take about 50-100 milliseconds under most circumstances.  Like any cloud service we are prone to network latency (both incoming and outgoing) but where we're running our Multi-tenant platform on AWS the issue could just be that you're further from the datacenter that Kinvey operates from?

As an aside:  The newest version of our iOS library (3.0 beta -- which is still a beta release) incorporates Realm as well as more advanced offline functionality.  

If anyone has any questions or any thoughts they want to share, please stop by our forums and give us a shout.  

Thanks,

**Comment by damien-kinvey** [comment]
Good morning everyone,

My name is Damien and I'm the Technical Support Team lead with Kinvey. I apologize for my slow response time on your forum issue Bizzi-Body. I've been in charge of handling your issue and it slipped my mind to transfer it to another member of my team, I've been out sick since Monday with either strep throat or the flu (I just got back from the doctor's a few moments ago).

http://devcenter.kinvey.com/ios/guides/datastore#operators -- The operator KCSIn is the one that you're after which is listed in the documentation with this example:

KCSQuery(onField: "colors", usingConditional: .KCSIn, forValue: ["orange", "black"])

B. With regards to speed, we are running on the Mean stack and in my experience simple inserts take about 50-100 milliseconds under most circumstances. Like any cloud service we are prone to network latency (both incoming and outgoing) but where we're running our Multi-tenant platform on AWS the issue could just be that you're further from the datacenter that Kinvey operates from?

As an aside: The newest version of our iOS library (3.0 beta -- which is still a beta release) incorporates Realm as well as more advanced offline functionality.

If anyone has any questions or any thoughts they want to share, please stop by our forums and give us a shout.

Thanks,

**Comment by Bizzi-Body** [comment]
@damien-kinvey - thanks for the response.  

I'm not sure that usingConditional will work.  Imagine I was looking to search for the text "ang" - which I would expect to return the entry holding "Orange".  

Incorporating Realm into your library gets you extra points!  (I am a Realm fanboy)  I will definitely keep an eye out for your r3 release.  Is there a mailing list I can put my email address on for notification?

**Comment by thelubin** [comment]
@damien-kinvey Happy to see Kinvey support writing here! A less steep price model would be welcome. :-)

**Comment by Bizzi-Body** [comment]
@thelubin :+1:

**Comment by damien-kinvey** [comment]
@Bizzi-Body -- To your question, I don't think that you can do that, as our queries anchor to the beginning of strings being searched.   I am not sure of a way around this other than returning all the fields and searching on the client side through the returned array.

I don't know if there's a way around that at the moment.

@mikkokam -- We are a much smaller company than parse (we're about 50 full time employees) so our libraries are updated very frequently, but much of the work that's been taking place recently has been around beta library development (which is now available on our dev center for javascript based libraries).  

@thelubin -- I was going to comment on here sooner, this thread has been being watched by our team internally, but we wanted to let people come to this organically rather than trying to directly influence discussion.   That said, my being sick for the earlier part of this week probably set me back a few days in terms of discussion.  

As for pricing I don't have any control over that, but I can pass along feedback to the people that do have control over it.   I believe that we are coming up with a different pricing sku that addresses developers that are pre-production in a better fashion, though I don't know when that will be debuted or how much it will be or what features it will have at this time.   

For now I will pass along the feedback I've had here to our product team / the management team as a whole.  

Beyond that if you guys are able to contribute any points to the Parse to Kinvey discussion that we have on our forums would be greatly appreciated by all of us here at Kinvey and anyone else migrating.  While I have a parse account and a simple app up and running, my knowledge of the inner workings of parse relational data, cloud code, and other things was lacking.  

Thanks,

**Comment by dron247** [comment]
@damien-kinvey will kinvey provide long shut-down period and/or opensource server like parse did in case of end of business?

**Comment by damien-kinvey** [comment]
@dron247 -- I can't say for certain.   At this point we're still growing as a company.  I can't specify anything about our financial situation (I honestly don't know enough about it, and even if I did I'm under an NDA), but I can say that no end-of-company planning has been made at this time...At least to my knowledge

Pretty much the entire company except for Marketing and sales is made up of engineers.  In the event that we were to shut down we would be sure to give some kind of solution to our customers that addressed the need to migrate their data.  At present we do allow people to export the data to JSON (similar to parse).   

What I can say is this.  I've been with the company over a year, and I'm optimistic I'll have my job for a lot longer at the rate that it's going.  

Thanks,
